x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/TransportVersionClusterStateUpgradeIT.java

```diff
@@ -12,70 +12,135 @@
 import org.elasticsearch.Version;
 import org.elasticsearch.client.Request;
 import org.elasticsearch.common.util.Maps;
+import org.elasticsearch.test.rest.ObjectPath;
 
-import java.io.IOException;
-import java.util.List;
 import java.util.Map;
-import java.util.stream.Collectors;
 
 import static org.hamcrest.Matchers.equalTo;
-import static org.hamcrest.Matchers.everyItem;
+import static org.hamcrest.Matchers.oneOf;
 
 public class TransportVersionClusterStateUpgradeIT extends AbstractUpgradeTestCase {
 
-    public void testReadsInferredTransportVersions() throws IOException {
-        assumeTrue("TransportVersion introduced in 8.8.0", UPGRADE_FROM_VERSION.before(Version.V_8_8_0));
-        assumeTrue(
-            "This only has visible effects when upgrading beyond 8.8.0",
-            TransportVersion.current().after(TransportVersions.V_8_8_0)
-        );
-        assumeTrue("Only runs on the mixed cluster", CLUSTER_TYPE == ClusterType.MIXED);
-        // if the master is not upgraded, and the secondary node is, then the cluster info from the secondary
-        // should have inferred transport versions in it
-        // rely on randomisation to hit this case at some point in testing
+    private static final Version VERSION_INTRODUCING_TRANSPORT_VERSIONS = Version.V_8_8_0;
+    private static final Version VERSION_INTRODUCING_NODES_VERSIONS = Version.V_8_11_0;
+    private static final TransportVersion FIRST_TRANSPORT_VERSION = TransportVersions.V_8_8_0;
+
+    public void testReadsInferredTransportVersions() throws Exception {
+        assertEquals(VERSION_INTRODUCING_TRANSPORT_VERSIONS.id(), FIRST_TRANSPORT_VERSION.id());
 
-        Request request = new Request("GET", "/_cluster/state/nodes");
-        Map<String, Object> masterResponse = entityAsMap(client().performRequest(request));
-        assumeFalse("Master needs to not know about transport versions", masterResponse.containsKey("transport_versions"));
+        // waitUntil because the versions fixup on upgrade happens in the background so may need a retry
+        assertTrue(waitUntil(() -> {
```

Comment:
Will the default timeout be enough here on slow test clusters?

Reply:
In current `main` we don't retry at all and I couldn't find any failures in CI history so I think this will be fine.